src/Storages/ObjectStorage/S3/Configuration.cpp

```diff
@@ -293,77 +302,35 @@ bool StorageS3Configuration::collectCredentials(ASTPtr maybe_credentials, S3::S3
     return true;
 }
 
-template <typename T>
-static std::optional<T> getFromPositionOrKeyValue(
-    const std::string & key,
-    const ASTs & args,
-    const std::unordered_map<std::string_view, size_t> & engine_args_to_idx,
-    const std::unordered_map<std::string, Field> & key_value_args)
+void StorageS3Configuration::fromDisk(const String & disk_name, ASTs & args, ContextPtr context, bool with_structure)
 {
-    if (auto arg_it = engine_args_to_idx.find(key); arg_it != engine_args_to_idx.end())
-        return checkAndGetLiteralArgument<T>(args[arg_it->second], key);
-
-    if (auto arg_it = key_value_args.find(key); arg_it != key_value_args.end())
-        return arg_it->second.safeGet<T>();
-
-    return std::nullopt;
-};
+    auto disk = context->getDisk(disk_name);
+    auto object_storage = disk->getObjectStorage();
+    const auto & s3_object_storage = assert_cast<const S3ObjectStorage &>(*object_storage);
+    s3_settings = std::make_unique<S3Settings>();
+    *s3_settings = s3_object_storage.getS3Settings();
 
-static std::unordered_map<std::string, Field> parseKeyValueArguments(const ASTs & function_args, ContextPtr context)
-{
-    std::unordered_map<std::string, Field> key_value_args;
-    for (const auto & arg : function_args)
+    ParseFromDiskResult parsing_result = parseFromDisk(args, with_structure, context);
     {
-        const auto * function_ast = arg->as<ASTFunction>();
-        if (!function_ast || function_ast->name != "equals")
-            continue;
-
-        auto * args_expr = assert_cast<ASTExpressionList *>(function_ast->arguments.get());
-        auto & children = args_expr->children;
-        if (children.size() != 2)
-        {
-            throw Exception(
-                ErrorCodes::BAD_ARGUMENTS,
-                "Key value argument is incorrect: expected 2 arguments, got {}",
-                children.size());
-        }
-
-        auto key_literal = evaluateConstantExpressionOrIdentifierAsLiteral(children[0], context);
-        auto value_literal = evaluateConstantExpressionOrIdentifierAsLiteral(children[1], context);
-
-        auto arg_name_value = key_literal->as<ASTLiteral>()->value;
-        if (arg_name_value.getType() != Field::Types::Which::String)
-            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expected string as credential name");
-
-        auto arg_name = arg_name_value.safeGet<String>();
-        auto arg_value = value_literal->as<ASTLiteral>()->value;
-
-        auto inserted = key_value_args.emplace(arg_name, arg_value).second;
-        if (!inserted)
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Duplicate key value argument: {}", arg_name);
+        String path = s3_object_storage.getURI().uri_str;
+        fs::path root = path;
+        fs::path suffix = parsing_result.path_suffix.value_or("");
+        setURL(S3::URI(String(root / suffix)));
```

Comment:
if `suffix` starts with `/`, then `root / suffix` will result in just `suffix` which we do not want to happen. Let's protect against that. Same for local configuration.